Test/paulmck/rcu.hpp

```diff
@@ -123,32 +104,23 @@ namespace std {
         }
     };
 
-    namespace details {
-	template<typename T, typename D = default_delete<T>>
-	class rcu_obj_base_ni: public rcu_head {
-	public:
-	    rcu_obj_base_ni(T *pi, D di = {}) {
-		p = pi;
-		d = di;
-	    }
-	    T *p;
-	    D d;
-	};
+    template<typename T, typename D = default_delete<T>>
+    void rcu_retire(std::unique_ptr<T, D> ptr)
+    {
+        typedef std::unique_ptr<T, D> ptr_type;
+        struct rcu_obj_base_ni :
+            public rcu_obj_base<rcu_obj_base_ni>,
+            public ptr_type {
+            rcu_obj_base_ni(ptr_type p) : ptr_type(std::move(p)) {}
+        };
+        (new rcu_obj_base_ni(std::move(ptr)))->retire();
```

Comment:
I agree, it isn't a good idea to allocate on the free path. However, I don't think that rcu_retire has any other option.